include/revng/Model/Type.h

```diff
@@ -0,0 +1,726 @@
+// Make model::Type derived types usable with UpcastablePointer
+template<>
+struct concrete_types_traits<model::Type> {
+  using type = std::tuple<model::PrimitiveType,
+                          model::EnumType,
+                          model::TypedefType,
+                          model::StructType,
+                          model::UnionType,
+                          model::FunctionPointerType>;
+};
+
+template<>
+struct concrete_types_traits<const model::Type> {
+  using type = std::tuple<const model::PrimitiveType,
+                          const model::EnumType,
+                          const model::TypedefType,
+                          const model::StructType,
+                          const model::UnionType,
+                          const model::FunctionPointerType>;
+};
+
+/// \brief Concept to identify all types that are derived from model::Type
+template<typename T>
+concept FlatCType = std::is_base_of_v<model::Type, T>;
+
+/// \brief Make all types derived from model::Type yaml-serializable like tuples
+template<FlatCType T>
+struct llvm::yaml::MappingTraits<T> : public TupleLikeMappingTraits<T> {};
+
+/// \brief Base class of model types used for LLVM-style RTTI
+struct model::Type {
+  // static constexpr const char *Tag = "!Invalid";
+  using Key = std::pair<model::TypeKind::Values, uint64_t>;
+
+  /// \brief The kind of a model::Type
+  TypeKind::Values Kind;
+
+  // TODO: we should switch to a sane UUID implementation
+  uint64_t ID;
+
+  /// \brief The name of this type. This is the unique key to identifiy a type.
+  std::string Name;
+
+public:
+  static bool classof(const Type *T) { return false; }
+
+  Key key() const { return Key{ Kind, ID }; }
+
+protected:
+  Type(TypeKind::Values TK, llvm::StringRef NameRef) : Type(TK, 0ULL, NameRef) {
+    auto Err = llvm::getRandomBytes(&ID, sizeof(ID));
+    if (Err)
+      ID = 0ULL;
+  }
+
+  Type(TypeKind::Values TK, uint64_t TheID, llvm::StringRef NameRef) :
+    Kind(TK), ID(TheID), Name(NameRef) {}
+
+public:
+  template<FlatCType T>
+  bool operator==(const T &Other) const {
+    return key() == Other.key();
+  }
+
+  template<FlatCType T>
+  bool operator<=(const T &Other) const {
+    return key() <= Other.key();
+  }
+
+protected:
+  bool verify() const debug_function;
+};
+INTROSPECTION_NS(model, Type, Kind, ID, Name);
+
+namespace model {
+
+using UpcastableType = UpcastablePointer<model::Type>;
+
+} // end namespace model
+
+template<>
+struct llvm::yaml::ScalarTraits<model::Type::Key>
+  : CompositeScalar<model::Type::Key, '#'> {};
+
+// Make UpcastablePointers to model types usable in KeyedObject containers
+template<>
+struct KeyedObjectTraits<model::UpcastableType> {
+  static model::Type::Key key(const model::UpcastableType &Val) {
+    return Val->key();
+  }
+  static model::UpcastableType fromKey(const model::Type::Key &K);
+
+private:
+  template<size_t I = 0>
+  static model::UpcastableType
+  makeFromID(model::TypeKind::Values Kind, uint64_t ID);
```

Comment:
Uhm, if you want to make this part of the trait make it generic with argument pack and squillo di lusso, otherwise I'd go for a regular function.

Reply:
And regular function it is.